### PR TITLE
[upath io manager] if an asset goes from unpartitioned to partitioned, delete the old storage file so the partitioned data can be stored

### PR DESCRIPTION
**Pull request by jamiedemaria**
## Summary & Motivation
The UPath I/O managers store `None` values. This presents a couple problems, but one of the most user facing issues is that if a user transitions an asset from non-partitioned to partitioned, they will get a FileExists error. For new users, especially users who are not introduced to I/O managers, this could be a pretty common experience. 

```python
@asset 
def my_asset():
   # does some work, but returns None
```
if they materialize this asset, the file `$DAGSTER_HOME/storage/my_asset` will be created.
If they later update `my_asset` to be partitioned
```python
@asset(partitions_def=DailyPartitionsDefinition(...))
def my_asset():
   # does some work, but returns None
```
and try to materialize, they will get the error 
```
FileExistsError: [Errno 17] File exists: '$DAGSTER_HOME/storage/my_asset'
```
since the I/O manager wants to store the partitioned data at `$DAGSTER_HOME/storage/my_asset/<partition_key>`.

The solution for this right now is to manually delete the `my_asset` file, or annotate the asset with the `-> None` return type. However, we could detect that the `my_asset` file exists, and delete it for the user in the UPath I/O manager. I think this is a safe thing to do since the user is materializing the asset anyway, so they would be overriding the contents of the file even if they hadn't transitioned it to a partitioned asset.

## How I Tested These Changes
new unit test. Ran it before making code changes and confirmed the FileExists error was raised. Error no longer raised with the new code changes

### CODE REVIEW

**Comment by jamiedemaria** [comment]
Current dependencies on/for this PR:
* `master`
  * **PR #19613** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19613?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/19613?utm_source=stack-comment).

**Comment by tacastillo** [comment]
you have no idea how excited I am about this!

QQ: Unsure if this is already an established path in IO managers, but what happens if:
1. user defines an asset and materializes it
2. adds a partition to it
3. tries to materialize a partition, but due to some force majeure, `handle_output` fails after deleting the file and before inserting the new one?

The existing asset would be deleted. This might be okay though because Dagster would also be communicating that the asset is in a Failed state, but wanted to pose the question.

File: python_modules/dagster/dagster/_core/storage/upath_io_manager.py [hunk 1]
```diff
@@ -425,6 +425,13 @@ def load_input(self, context: InputContext) -> Union[Any, Dict[str, Any]]:
 
                 return self._load_multiple_inputs(context)
 
+    def _handle_transition_to_partitioned_asset(self, path: "UPath"):
+        # if the asset was previously non-partitioned, path will be a file, when it should
+        # be a directory for the partitioned asset. Delete the file, so that it can be made
+        # into a directory
+        if path.exists() and path.is_file():
+            path.unlink()
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/storage/upath_io_manager.py:433]: * I think we should `context.log.warn` when this happens for clarity to user
* do we need to worry about this happening concurrently? Upath io manager has that coroutine support for partitions for somethings (i forget if its only load input or also handle output). I think you can do `missing_ok=True` to make it so this is safe if theres a race and we try to unlink something that is gone

Comment (jamiedemaria) [reply python_modules/dagster/dagster/_core/storage/upath_io_manager.py:433]: I'll add the `missing_ok=True` just to be safe. two materializations of the same asset that both detect the existing file, but one deletes it before the other tries to delete is unlikely but possible.

**Comment by jamiedemaria** [comment]
@tacastillo for you question- i \~think\~ this should be ok - we run the same risk with the database i/o managers since we manually remove the relevant rows before inserting the new rows in. But this risk doesn't exist in the upath i/o managers now, so we are adding it in. I personally don't think it make this a no-go, but if other's are really concerned about it we could think of solutions (maybe copying the data to tmp file and then restoring it to the main file if the write fails?)

**Comment by tacastillo** [comment]
@jamiedemaria yeah agreed this isn't a blocker. cool cool! yeah and I don't imagine the common uses of this IO manager to ever hit this situation and find it much less risky than the db io sitch.

**Review by tacastillo — APPROVED** [review]
all hail the io manager manager

**Comment by jamiedemaria** [comment]
@alangenfeld  did you plan on taking another pass at this?

File: python_modules/dagster/dagster/_core/storage/upath_io_manager.py [hunk 2]
```diff
@@ -425,6 +425,18 @@ def load_input(self, context: InputContext) -> Union[Any, Dict[str, Any]]:
...
                 return self._load_multiple_inputs(context)
 
+    def _handle_transition_to_partitioned_asset(self, context: OutputContext, path: "UPath"):
+        # if the asset was previously non-partitioned, path will be a file, when it should
+        # be a directory for the partitioned asset. Delete the file, so that it can be made
+        # into a directory
+        if path.exists() and path.is_file():
+            context.log.warn(
+                f"Non-partitioned asset {context.asset_key} was materialized to"
+                " file {path}. Materializing {context.asset_key} as a partitioned asset,"
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/storage/upath_io_manager.py:435]: not f-string-ed  `f"`

File: python_modules/dagster/dagster/_core/storage/upath_io_manager.py [hunk 3]
```diff
@@ -425,6 +425,18 @@ def load_input(self, context: InputContext) -> Union[Any, Dict[str, Any]]:
...
+                f" will delete this file, and create a directory, {path} containing partitioned data files."
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/storage/upath_io_manager.py:436]: the tense here reads a little weird to me, this is going to happen so maybe change it up. What pops in my head is something like: 

"Found file at {path} believed to correspond with previously non-partitioned version of {asset}. Removing file and replacing with directory for partitioned data files"

**Review by alangenfeld — APPROVED** [review]
ah my B, accepted PRs drop out of my review queue. Take one last past on the message and this should be good

**Pull request merged**